### PR TITLE
Fixes Telescope Issue #1936, added Cache-Control microservice

### PR DESCRIPTION
**Pull request by Metropass**
This is a middleware service that addresses the needs for Cache-Control Levels.

### Arguments:
- `res` : The object that will be assigned our Cache-Control levels

#### Cache Durations:
`never(res)` - The `never()` function doesn't not allow caching at all, `max-age` is set as 0, along with `no-store`. 
`duration(time, res)` - Allows a user to set a custom time for `max-age`. 
`forever(res)` - allows caching for up to 1 year, and sets the item as immutable. 

#### Freshness, and Staleness
It's possible to set your Response to a certain freshness, or how long the response can be stale for.

`freshFor(time, res)` - Sets the amount of time a response could be fresh for.
`staleFor(time, res)` - Sets the amount of time a response could be stale for.

When calling our Cache Control functions, this is how you call the functions/
For example, 

```js
cache().forever(res); // Sets max-age to 1 year forever
cache().duration('1m', res); // Sets max-age to 1 month
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "express-pino-logger": "^6.0.0",
     "helmet": "4.4.1",
     "http-errors": "^1.8.0",
+    "ioredis-mock": "^5.4.1",
```
Comment (manekenpix) [inline package.json:31]: This shouldn't be part of this PR

**Review by manekenpix — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Metropass) [reply package.json]: oh my, thanks for the heads up. it's removed now

Comment (chrispinkney) [reply package.json]: But now the tests are failing because it was removed haha

Comment (Metropass) [reply package.json]: Yeah lol, I just fixed it along with fixing the misspelt function names

File: test.js [hunk 2]
```diff
@@ -15,6 +15,8 @@ const {
   hash,
   createError,
 } = require('./src');
+const { setHttpCacheHeaders } = require('./src/middleware');
+const { request } = require('http');
```
Comment (manekenpix) [inline test.js:19]: Where is this use?

File: src/middleware.js [hunk 3]
```diff
@@ -112,6 +112,82 @@ function errorHandler(err, req, res, next) {
...
+        createError(
+          500,
+          "Unexpected Error: ${error}, check to see if you've made a mistake when defining options."
+        )
+      );
+    }
+  } else {
+    switch (levelValue) {
+      case LEVEL1:
+        res.set('Cache-Control', 'public, max-age=0');
```
Comment (chrispinkney) [inline src/middleware.js:159]: Is there any way this could (_should?_) allow the dev to specify their own `max-age`? Kind of like how you can pass options in your example?

I'm asking mostly because I'm not sure if this is functionality is something that **should** be added in, or if we should only allow what is defined, no more, no less.

Comment (chrispinkney) [reply src/middleware.js:159]: Also the jump from 300 to 2592000 is pretty big, should it be? (genuine question, I'm not sure)

File: test.js (hunk 2 above)
Comment (Metropass) [reply test.js:19]: Line 18 is the import for my module, I don't use Line 19, I'll remove that

File: src/middleware.js (hunk 3 above)
Comment (Metropass) [reply src/middleware.js:159]: That's a good question, I originally designed it so that we could set pre-defined limits on how long we could cache our data. 

However you make a good point, cause maybe we might need something to be cached for a specific time. 

I'll add in an option for that, just in case, Thanks 😄

Comment (chrispinkney) [reply src/middleware.js:159]: Np!

File: src/middleware.js [hunk 4]
```diff
@@ -112,6 +112,95 @@ function errorHandler(err, req, res, next) {
...
+        break;
+      case LEVEL2:
+        res.set('Cache-Control', 'public, max-age=300');
+        break;
+      case LEVEL3:
+        res.set('Cache-Control', 'public, max-age=2592000');
+        break;
+      case LEVEL4:
+        res.set('Cache-Control', 'public, max-age=31536000, immutable');
```
Comment (manekenpix) [inline src/middleware.js:169]: Would it make sense to change those numbers to something meaningful like
```js
const oneMonth  = 1000 * 60 * 60 * 24 * 30; // instead of 2592000
res.set('Cache-Control', `public, max=${oneMonth}`);
```

Comment (humphd) [reply src/middleware.js:169]: https://www.npmjs.com/package/ms does this

**Review by humphd — CHANGES_REQUESTED** [review]
I'm not sure that this solves a real problem.  You can already call `res.set(...)` to set headers.  Is the idea to provide some default versions?  If so, the `LEVEL1` style terminology needs to be more explicit.  You could do:

```js
const { cache } = require('@senecacdot/satellite');
...
// Cache content forever (e.g., 1 year)
service.get('/', cache.forever(), (res, req) => {...});

// Never Cache
service.get('/', cache.never(), (res, req) => {...});

// Cache for some specific time, use https://www.npmjs.com/package/ms to parse
service.get('/', cache.duration('7 days'), (res, req) => {...});
```

File: test.js [hunk 5]
```diff
@@ -779,45 +781,89 @@ describe('Satellite()', () => {
       });
       helmetService.start(port, async () => {
         const res = await fetch(`${url}/always-200`);
+        // console.log(res);
```
Comment (manekenpix) [inline test.js:784]: I think you can delete this

File: src/middleware.js [hunk 6]
```diff
@@ -1,6 +1,8 @@
 const jwt = require('express-jwt');
 const createError = require('http-errors');
 const logger = require('./logger');
+const ms = require('ms');
+const express = require('express');
```
Comment (humphd) [inline src/middleware.js:5]: This doesn't seem to be used in this file, you can remove.

File: src/middleware.js [hunk 7]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
...
+    res.headers.set(`Cache-Control`, `public, max-age=${t}, immutable`);
+  }; // Forever Function
+
+  /*
+no-store is needed with max-age=0
+we don't want to serve stale resources when Telescope is down.
+see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#requiring_revalidation
+*/
+  this.never = (res) => {
+    res.headers.set('Cache-Control', 'public, no-store, max-age=0');
```
Comment (humphd) [inline src/middleware.js:129]: Let's not include `public` in case it isn't. I think you can use `Cache-Control: no-store, max-age=0`

File: src/middleware.js [hunk 8]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
   }
 }
 
+function cache(res) {
```
Comment (humphd) [inline src/middleware.js:117]: I think `cache` should be an object vs a function, so you'd use it like this:

```js
app.get('/', cache.forever(), (res, res) => {...});
```

That means:

```js
const cache = {
  forever() {
    const t = ms('1y');

    return function(req, res, next) {
      res.set('Cache-Control', `public, max-age=${t}, immutable`);
      next();
    };
  },

  never(req, res, next) {
    ...
  }, 

  ...
};
```

Also, note the signature `(req, res, next)` that all Express middleware needs, and you have to call `next()` when done.  Last, `res.set()` vs. `res.headers.set()`.

File: src/middleware.js [hunk 9]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
...
+  /*
+no-store is needed with max-age=0
+we don't want to serve stale resources when Telescope is down.
+see: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#requiring_revalidation
+*/
+  this.never = (res) => {
+    res.headers.set('Cache-Control', 'public, no-store, max-age=0');
+  };
+
+  this.duration = (time, res) => {
```
Comment (humphd) [inline src/middleware.js:132]:
```js
const cache = {
  duration(time) {
    const t = ms(time);

    return function(req, res, next) {
      res.set('Cache-Control', `public, max-age=${t}`);
      next();
    };
  },
  ..

}
```

Used like this:

```js
app.get('/', cache.duration('1d'), (req, res) => {...});
```

File: src/middleware.js [hunk 10]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
   }
 }
 
+function cache(res) {
+  this.forever = (res) => {
+    const t = ms('1y');
+    res.headers.set(`Cache-Control`, `public, max-age=${t}, immutable`);
+  }; // Forever Function
+
+  /*
```
Comment (humphd) [inline src/middleware.js:123]: This comment should either get included in the code somehow or removed.

File: src/middleware.js [hunk 11]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
...
+  this.never = (res) => {
+    res.headers.set('Cache-Control', 'public, no-store, max-age=0');
+  };
+
+  this.duration = (time, res) => {
+    const t = ms(time);
+    res.headers.set('Cache-Control', `public, max-age=${t}`);
+  };
+
+  this.staleFor = (time, res) => {
```
Comment (humphd) [inline src/middleware.js:137]: What does this one do?

File: src/middleware.js [hunk 12]
```diff
@@ -112,6 +114,62 @@ function errorHandler(err, req, res, next) {
...
+      next(
+        createError(
+          428,
+          'You do not have a Cache-Control header, this function is used for attributes that contain the header. '
+        )
+      );
+    }
+  };
+
+  this.freshFor = (time, res) => {
```
Comment (humphd) [inline src/middleware.js:151]: What about this one?

File: test.js [hunk 13]
```diff
@@ -779,45 +781,89 @@ describe('Satellite()', () => {
...
+  describe('cache', () => {
+    test('forever() should set the max-age to 1y, and add the immutable tag', (done) => {
+      const service = createSatelliteInstance({
+        name: 'header-test',
+        port,
+      });
+      service.start(port, async () => {
+        const res = await fetch(`${url}/always-200`);
+        expect(res.ok).toBe(true);
+        cache().forever(res);
```
Comment (humphd) [inline test.js:801]: You have to add the middleware to a new route.  Something like this:

```js
    const service = createSatelliteInstance({
      name: 'test',
    });
    const router = service.router;
    router.get('/cached-forever', cache.forever(), (req, res) => res.json({ hello: 'public' }));
```

Now you can try and `fetch()` that and check the headers:

```js
    service.start(port, async () => {
      let res = await fetch(`${url}/cached-forever`);
      expect(res.ok).toBe(true);
      expect(res.headers.get('Cache-Control')).toBe('public, max-age=31557600000, immutable');
      ...
```

**Review by humphd — CHANGES_REQUESTED** [review]

**Comment by humphd** [comment]
Let me know when you want this reviewed again, it looks like you're still working?

**Comment by Metropass** [comment]
> Let me know when you want this reviewed again, it looks like you're still working?

@humphd  Yup! I am. Quite a lot of progress though, I got nginx to set our Cache Control headers, along with Read/Write to the config file by class. If by any chance, we would need to add anything in our nginx config dynamically, we could do so now, along with setting our Caches to their respective levels as defined above. 

Eg: finding all the values for `add_header` within the `http` portion of the config
Screenshot_413
`value` is a variable passed, and it's the block within a config (in the pic, `value` is set to `http`)

it also supports adding in new blocks if we ever need it:
Screenshot_414
`prop` is the name of the new block that we would like to create.
